File: hooks/scripts/hooks/session-init-helpers.ts
// session-init helper functions: shadowed skills cleanup + agent team detection
import * as fs from "node:fs";
import * as path from "node:path";
import * as os from "node:os";

interface ShadowedCleanupResult {
  restored: string[];
  skipped: string[];
  kept: string[];
}

interface TeamInfo {
  teamName: string;
  memberCount: number;
}

/** One-time cleanup for orphaned .shadowed/ dirs from disabled skill-dedup hook (Issue #422) */
export function cleanupOrphanedShadowedSkills(): ShadowedCleanupResult {
  const shadowedDir = path.join(process.cwd(), ".claude", "skills", ".shadowed");
  if (!fs.existsSync(shadowedDir)) return { restored: [], skipped: [], kept: [] };

  const skillsDir = path.join(process.cwd(), ".claude", "skills");
  const restored: string[] = [];
  const skipped: string[] = [];
  const kept: string[] = [];

  try {
    const entries = fs.readdirSync(shadowedDir, { withFileTypes: true });
    for (const entry of entries) {
      if (!entry.isDirectory()) continue;
      const src = path.join(shadowedDir, entry.name);
      const dest = path.join(skillsDir, entry.name);
      try {
        if (!fs.existsSync(dest)) {
          fs.renameSync(src, dest);
          restored.push(entry.name);
        } else {
          const orphanedSkill = path.join(src, "SKILL.md");
          const localSkill = path.join(dest, "SKILL.md");
          if (fs.existsSync(orphanedSkill) && fs.existsSync(localSkill)) {
            if (fs.readFileSync(orphanedSkill, "utf8") === fs.readFileSync(localSkill, "utf8")) {
              fs.rmSync(src, { recursive: true, force: true });
              skipped.push(entry.name);
            } else {
              kept.push(entry.name); // Content differs — keep for manual review
            }
          } else {
            fs.rmSync(src, { recursive: true, force: true });
            skipped.push(entry.name);
          }
        }
      } catch (err) {
        process.stderr.write(`[session-init] Failed to process "${entry.name}": ${(err as Error).message}\n`);
      }
    }

    const manifestFile = path.join(shadowedDir, ".dedup-manifest.json");
    if (fs.existsSync(manifestFile)) fs.unlinkSync(manifestFile);
    const remaining = fs.readdirSync(shadowedDir);
    if (remaining.length === 0) fs.rmdirSync(shadowedDir);
  } catch (err) {
    process.stderr.write(`[session-init] Shadowed cleanup error: ${(err as Error).message}\n`);
  }

  return { restored, skipped, kept };
}

/** Detect if session is running inside an Agent Team by scanning ~/.claude/teams/ */
export function detectAgentTeam(): TeamInfo | null {
  try {
    const teamsDir = path.join(os.homedir(), ".claude", "teams");
    if (!fs.existsSync(teamsDir)) return null;

    for (const entry of fs.readdirSync(teamsDir, { withFileTypes: true })) {
      if (!entry.isDirectory()) continue;
      const configPath = path.join(teamsDir, entry.name, "config.json");
      if (!fs.existsSync(configPath)) continue;
      try {
        const config = JSON.parse(fs.readFileSync(configPath, "utf8")) as { members?: unknown[] };
        if (config.members && config.members.length > 0) {
          return { teamName: entry.name, memberCount: config.members.length };
        }
      } catch {}
    }
    return null;
  } catch {
    return null;
  }
}

File: hooks/scripts/cmd/session_init_helpers.go
package cmd

import (
	"encoding/json"
	"fmt"
	"os"
	"path/filepath"

	"solon-hooks/internal/config"
)

// writeEnvForced writes an env var to envFile even if value is empty string.
// Unlike config.WriteEnv which skips zero/empty values.
func writeEnvForced(envFile, key string, value interface{}) {
	if envFile == "" {
		return
	}
	escaped := config.EscapeShellValue(fmt.Sprintf("%v", value))
	line := fmt.Sprintf("export %s=\"%s\"\n", key, escaped)
	f, err := os.OpenFile(envFile, os.O_APPEND|os.O_CREATE|os.O_WRONLY, 0644)
	if err != nil {
		return
	}
	defer f.Close()
	_, _ = f.WriteString(line)
}

// shadowedCleanupResult holds results of orphaned .shadowed/ dir cleanup.
type shadowedCleanupResult struct {
	Restored []string
	Skipped  []string
	Kept     []string
}

// teamInfo holds detected agent team info.
type teamInfo struct {
	TeamName    string
	MemberCount int
}

// cleanupOrphanedShadowedSkills restores skills from an orphaned .shadowed/ dir (Issue #422).
func cleanupOrphanedShadowedSkills() shadowedCleanupResult {
	cwd, _ := os.Getwd()
	shadowedDir := filepath.Join(cwd, ".claude", "skills", ".shadowed")
	skillsDir := filepath.Join(cwd, ".claude", "skills")

	if _, err := os.Stat(shadowedDir); os.IsNotExist(err) {
		return shadowedCleanupResult{}
	}

	var result shadowedCleanupResult
	entries, err := os.ReadDir(shadowedDir)
	if err != nil {
		return result
	}

	for _, entry := range entries {
		if !entry.IsDir() {
			continue
		}
		src := filepath.Join(shadowedDir, entry.Name())
		dest := filepath.Join(skillsDir, entry.Name())

		if _, err := os.Stat(dest); os.IsNotExist(err) {
			// Destination doesn't exist — restore
			if err := os.Rename(src, dest); err == nil {
				result.Restored = append(result.Restored, entry.Name())
			} else {
				os.Stderr.WriteString("[session-init] Failed to restore \"" + entry.Name() + "\": " + err.Error() + "\n")
			}
			continue
		}

		// Destination exists — compare SKILL.md
		orphanSkill := filepath.Join(src, "SKILL.md")
		localSkill := filepath.Join(dest, "SKILL.md")
		orphanData, oErr := os.ReadFile(orphanSkill)
		localData, lErr := os.ReadFile(localSkill)

		if oErr == nil && lErr == nil && string(orphanData) == string(localData) {
			// Identical — safe to remove
			os.RemoveAll(src)
			result.Skipped = append(result.Skipped, entry.Name())
		} else if oErr != nil || lErr != nil {
			// Missing SKILL.md — remove
			os.RemoveAll(src)
			result.Skipped = append(result.Skipped, entry.Name())
		} else {
			// Content differs — keep for manual review
			result.Kept = append(result.Kept, entry.Name())
		}
	}

	// Clean up manifest and dir if empty
	manifestFile := filepath.Join(shadowedDir, ".dedup-manifest.json")
	os.Remove(manifestFile)
	remaining, _ := os.ReadDir(shadowedDir)
	if len(remaining) == 0 {
		os.Remove(shadowedDir)
	}

	return result
}

// detectAgentTeam scans ~/.claude/teams/ for an active team config.
func detectAgentTeam() *teamInfo {
	home, err := os.UserHomeDir()
	if err != nil {
		return nil
	}
	teamsDir := filepath.Join(home, ".claude", "teams")
	entries, err := os.ReadDir(teamsDir)
	if err != nil {
		return nil
	}
	for _, entry := range entries {
		if !entry.IsDir() {
			continue
		}
		configPath := filepath.Join(teamsDir, entry.Name(), "config.json")
		data, err := os.ReadFile(configPath)
		if err != nil {
			continue
		}
		var cfg struct {
			Members []interface{} `json:"members"`
		}
		if err := json.Unmarshal(data, &cfg); err != nil {
			continue
		}
		if len(cfg.Members) > 0 {
			return &teamInfo{TeamName: entry.Name(), MemberCount: len(cfg.Members)}
		}
	}
	return nil
}
